src/EditorFeatures/Core/Implementation/Structure/AbstractStructureTaggerProvider.cs

```diff
@@ -252,13 +200,92 @@ protected sealed override ITaggerEventSource CreateEventSource(ITextView textVie
+
+        /// <summary>
+        /// Given a <see cref="StructureTag"/>, creates an ITextBuffer with the content to display 
+        /// in the tooltip.
+        /// </summary>
+        protected ITextBuffer CreateElisionBufferForTagTooltip(StructureTag tag)
+        {
+            // Remove any starting whitespace.
+            var span = TrimStartingNewlines(new SnapshotSpan(tag.Snapshot, tag.CollapsedHintFormSpan));
+
+            // Trim the length if it's too long.
+            var shortSpan = span;
+            if (span.Length > MaxPreviewText)
+            {
+                shortSpan = ComputeShortSpan(span);
+            }
+
+            // Create an elision buffer for that span, also trimming the
+            // leading whitespace.
+            var elisionBuffer = CreateElisionBufferWithoutIndentation(shortSpan);
+            var finalBuffer = elisionBuffer;
+
+            // If we trimmed the length, then make a projection buffer that 
+            // has the above elision buffer and follows it with "..."
+            if (span.Length != shortSpan.Length)
+            {
+                finalBuffer = CreateTrimmedProjectionBuffer(elisionBuffer);
+            }
+
+            return finalBuffer;
+        }
+
+        private ITextBuffer CreateTrimmedProjectionBuffer(ITextBuffer elisionBuffer)
+        {
+            // The elision buffer is too long.  We've already trimmed it, but now we want to add
+            // a "..." to it.  We do that by creating a projection of both the elision buffer and
+            // a new text buffer wrapping the ellipsis.
+            var elisionSpan = elisionBuffer.CurrentSnapshot.GetFullSpan();
+
+            var sourceSpans = new List<object>()
+                {
+                    elisionSpan.Snapshot.CreateTrackingSpan(elisionSpan, SpanTrackingMode.EdgeExclusive),
+                    "..."
+                };
+
+            var projectionBuffer = ProjectionBufferFactoryService.CreateProjectionBuffer(
+                projectionEditResolver: null,
+                sourceSpans: sourceSpans,
+                options: ProjectionBufferOptions.None);
+
+            return projectionBuffer;
+        }
+
+        private static SnapshotSpan ComputeShortSpan(SnapshotSpan span)
+        {
+            var endIndex = span.Start + MaxPreviewText;
+            var line = span.Snapshot.GetLineFromPosition(endIndex);
+
+            return new SnapshotSpan(span.Snapshot, Span.FromBounds(span.Start, line.EndIncludingLineBreak));
+        }
+
+        private static SnapshotSpan TrimStartingNewlines(SnapshotSpan span)
+        {
+            while (span.Length > 1 && char.IsWhiteSpace(span.Snapshot[span.Start]))
+            {
+                span = new SnapshotSpan(span.Snapshot, span.Start + 1, span.Length - 1);
+            }
+
+            return span;
+        }
+
+        private ITextBuffer CreateElisionBufferWithoutIndentation(
+            SnapshotSpan shortHintSpan)
```

Comment:
inline this method?

Reply:
Unless you feel strongly going to pass -- the caller is already pretty non-trivial to understand as it is.